demos/fleet_provisioning/fleet_provisioning_with_csr/fleet_provisioning_csr_demo.c

```diff
@@ -0,0 +1,874 @@
+        }
+    }
+
+    return status;
+}
+/*-----------------------------------------------------------*/
+
+static bool saveCertificate( const char * pCertificate,
+                             size_t certificateLength )
+{
+    FILE * file;
+    bool status = false;
+
+    file = fopen( PROVISIONING_CERT_PATH, "wb" );
+
+    if( file == NULL )
+    {
+        LogError( ( "Failed opening PROVISIONING_CERT_PATH for writing." ) );
+    }
+    else
+    {
+        size_t written;
+        written = fwrite( ( const void * ) pCertificate, 1U, certificateLength, file );
+        fclose( file );
+
+        if( written != certificateLength )
+        {
+            LogError( ( "Failed writing certificate to file." ) );
+        }
+        else
+        {
+            status = true;
+        }
+    }
+
+    return status;
+}
+/*-----------------------------------------------------------*/
+
+/* This example uses a single application task, which shows that how to use
+ * the Fleet Provisioning library to generate and validate AWS IoT Fleet
+ * Provisioning MQTT topics, and use the coreMQTT library to communicate with
+ * the AWS IoT Fleet Provisioning APIs. */
+int main( int argc,
+          char ** argv )
+{
+    bool status = false;
+    int exitStatus = EXIT_FAILURE;
+    char serial[ 10 ];
+    size_t serialLength = 10;
+    /* Buffer for holding the CSR. */
+    char csr[ NETWORK_BUFFER_SIZE ] = { 0 };
+    size_t csrLength = 0;
+    /* Buffer for holding received certificate until it is saved. */
+    char certificate[ CERT_BUFFER_LENGTH ];
+    size_t certificateLength;
+    /* Buffer for holding the certificate ID. */
+    char certificateId[ CERT_ID_BUFFER_LENGTH ];
+    size_t certificateIdLength;
+    /* Buffer for holding the certificate ownership token. */
+    char ownershipToken[ OWNERSHIP_TOKEN_BUFFER_LENGTH ];
+    size_t ownershipTokenLength;
+    bool connectionEstablished = false;
+    size_t i;
+    int demoRunCount = 0;
+
+    /* Silence compiler warnings about unused variables. */
+    ( void ) argc;
+    ( void ) argv;
+
+    do
+    {
+        /* Initialize the buffer lengths to their max lengths. */
+        certificateLength = CERT_BUFFER_LENGTH;
+        certificateIdLength = CERT_ID_BUFFER_LENGTH;
+        ownershipTokenLength = OWNERSHIP_TOKEN_BUFFER_LENGTH;
+
+        /* The demo template we use has a serial number as a parameter.
+         * For this demo we use a random 10 digit number. */
+        for( i = 0; i < serialLength; i++ )
+        {
+            serial[ i ] = '0' + ( rand() % 10 );
+        }
+
+        /* Read the CSR into the CSR buffer. */
+        status = getCsr( csr, NETWORK_BUFFER_SIZE, &csrLength );
+
+        if( status == true )
+        {
+            /* Create the document containing the CSR to publish to the
+             * CreateCertificateFromCsr API. */
+            status = generateCsrRequest( payloadBuffer,
```

Comment:
Would be better to call this function just before the Publish operation to CreateCertificateFromCsr API